File: P344ReverseString/reverse_string.cxx
class Solution {
public:
    /*
     * @brief helper function for rec 
     * 
     * @param s input vector of characters
     * @param pos current position/index used in recursion
     */
    void reverse_string_recursive(std::vector<char> &s, int32_t pos)
    {
        if (pos >= (s.size() / 2)) {
            return;
        }

        reverse_string_recursive(s, pos + 1);

        int lo = pos;
        int hi = (s.size() - 1) - pos;

        std::swap(s[lo], s[hi]);

        return;
    }


    /*
     * @brief reverses a string using recursion
     * 
     * @param s input vector of characters
     */
    void reverseString(vector<char>& s) {        
        if (s.size() == 0) {
            return;
        }

        reverse_string_recursive(s, 0);
    }
};


class Solution {
public:
     /*
      * @brief reverses a string iteratively
      * 
      * @param s
      */
    void reverseString(vector<char>& s) {
        size_t left = 0;
        size_t right = s.size() - 1;
        
        const size_t mid = s.size() / 2;
        
        for (size_t i = 0; i < mid; ++i) {
            char tmp = s[left];
            s[left]  = s[right];
            s[right] = tmp;
            
            ++left;
            --right;
        }
    }
};

File: P344ReverseString/reverse_string.py
class Solution:
    def reverseString(self, s: List[str]) -> None:
        """
        Reverses a string represented as a list of characters
        in-place. No function methods from the list class such as reverse() 
        are used.
        
        This approach uses two pointers (not literal pointers in the 
        C/C++ sense), one for the front (lo), and one that iterates 
        from the end (hi). The characters at positions lo and hi are
        swapped, the lo pointer is incremeneted to the next character,
        while the hi pointer is decremented to the character before the
        current hi position. The algorithm terminates once lo is equal to hi
        (in the case of a string with odd length) or greater than hi (in the
        case of a string of even length). The time complexity is O(n) and 
        the space complexity is constant O(1).
        
        Consider the example "hello, world". 
                               swap characters        move pointers
    
        "hello, world"   -->   "dello, worlh"  -->   "dello, worlh" 
         ^          ^           ^          ^           ^        ^
        lo          hi         lo          hi         lo        hi

        "dello, worlh"   -->   "dlllo, woreh"  -->   "dlllo, woreh" 
          ^        ^             ^        ^             ^      ^
          lo       hi           lo        hi           lo      hi
        
        "dlllo, woreh"   -->   "dlrlo, woleh"  -->   "dlrlo, woleh" 
           ^      ^               ^      ^               ^    ^   
          lo      hi             lo      hi             lo    hi
        
        "dlrlo, woleh"   -->   "dlroo, wlleh"  -->   "dlroo, wlleh" 
            ^    ^                 ^    ^                 ^  ^   
           lo    hi               lo    hi               lo  hi
        
        "dlroo, wlleh"   -->   "dlrow, olleh"  -->   "dlrow, olleh" 
             ^  ^                   ^  ^                   ^^   
            lo  hi                 lo  hi                 lohi
        
        "dlrow, olleh"   -->   "dlrow ,olleh"  -->   "dlrow, olleh" 
              ^^                     ^^                    ^^   
             lohi                   lohi                  hilo <- break (hi < lo)

        Parameters
        ----------
        s: List
            List of characters to be reversed in-place
    
        Examples
        --------
        Input: ["H","a","n","n","a","h"]
        Output: ["h","a","n","n","a","H"]

        Input: ["w","i","n","t","e","r"]
        Output ["r","e","t","n","i","w"]
        """
        lo = 0
        hi = len(s) - 1
        
        while lo < hi:
            s[lo], s[hi] = s[hi], s[lo]
            
            lo += 1
            hi -= 1